api/v1alpha1/restore_types.go

```diff
@@ -26,7 +26,9 @@ import (
 // one you need to define an existing PVC.
 type RestoreSpec struct {
 	// Backend contains the backend information
-	Backend       *Backend       `json:"backend,omitempty"`
+	Backend *Backend `json:"backend,omitempty"`
+	// FIXME(mw): does not work
+	// +kubebuilder:validation:required
```

Comment:
In https://book.kubebuilder.io/reference/markers/crd-validation.html it is uppercased, try this?
```suggestion
    // +kubebuilder:validation:Required
```
or `// +kubebuilder:validation:Required=true`

Reply:
doesn't work unfortunately. The uppercase version I've tried already and providing `=true` yields an error:
```
» make manifests
/Users/michael/code/go-workspace/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
/Users/michael/code/vshn/k8up/api/v1alpha1/restore_types.go:31:2: extra arguments provided: "true" (at <input>:1:1)
/Users/michael/code/vshn/k8up/api/v1alpha1/restore_types.go:31:2: extra arguments provided: "true" (at <input>:1:1)
/Users/michael/code/vshn/k8up/api/v1alpha1/restore_types.go:31:2: extra arguments provided: "true" (at <input>:1:1)
/Users/michael/code/vshn/k8up/api/v1alpha1/restore_types.go:31:2: extra arguments provided: "true" (at <input>:1:1)
```

Reply:
However, I was just quickly looking into the code and looked into the structs. So far I cannot see a reason why they have to be pointers at all. If it were
```go
type RestoreSpec struct {
	// Backend contains the backend information
	Backend       Backend       `json:"backend,omitempty"`...
}
```
then the Go deserializer will just fill the Backend field with an empty Backend struct. Now I don't want to question the API, but it would be nice if there's almost no pointers.

Reply:
The backend shouldn't be required for every CRD there can be defaults provided. That's also the main reasoning behind using pointers everywhere. We can explicitly check if a value was set or not. But on the other hand we HAVE to check if a value was set or not.

Examples in https://book.kubebuilder.io/cronjob-tutorial/api-design.html#designing-an-api use pointers for almost everything due to this.

Reply:
yeah pointer vs non-pointer .. classic Go discussion ;) The optional part is quite significant though, as for empty structs it's rather ugly to check whether they've been set or not. 

@Kidswiss what's your opinion on the validation part? i.e. marking everything optional and specifics required (or vice-versa)?

Reply:
The site you linked is just using int32 pointers for API compatibility reasons, but not the structs (e.g. `JobTemplate batchv1beta1.JobTemplateSpec 'json:"jobTemplate"'))`. I don't see any mentions regarding struct pointers...

But I don't want to involve myself too much here atm, I was just helping out Michael for the generation problem :) If you have reasons to do it that way, I'm perfectly fine with it :)

Reply:
@mweibel sounds reasonable, that should help with the endless nil checking for certain things.

@ccremer the pointers for the structs were used for the same reason as for the basic types: checking if it was explicitly set or not. Example is the logic to determine what backend is actually used.